test: lifespan tests load the developer's repo-root .env into os.environ — session-tracing validator tests fail on dev machines

## Symptom

Four session-tracing config-validation tests fail in **full-suite runs only** on a dev machine whose repo checkout has a root `.env` containing `LANGFUSE_PUBLIC_KEY` / `LANGFUSE_SECRET_KEY`. Each passes when its file runs alone, and CI is green (runners have no `.env`):

- `tests/test_session_tracing.py::TestConfigConstraints::test_langfuse_validator_keys_missing`
- `tests/test_session_tracing.py::TestConfigSaveValidationAndRollback::test_cli_config_set_validation_error`
- `tests/test_session_tracing.py::TestConfigSaveValidationAndRollback::test_web_api_patch_rollback[asyncio]`
- `tests/test_session_tracing.py::TestConfigSaveValidationAndRollback::test_mcp_config_rollback[asyncio]`

All four expect the `SessionTraceConfig` "requires both langfuse_public_key and langfuse_secret_key" validator to fire when no keys are configured; instead the operation succeeds:

```
E   AssertionError: assert 'Failed to persist config' in 'Set session_trace.langfuse_enabled = True (persisted to config.json)'
```

Observed on clean `origin/main @ cd0ae767` (surfaced while verifying #1506 / PR #1507 full runs, unrelated to that change).

## Minimal repro

Reproduces on any machine — fabricate the `.env` with fake keys at the repo root:

```bash
printf 'LANGFUSE_PUBLIC_KEY=pk-fake\nLANGFUSE_SECRET_KEY=sk-fake\n' > .env
uv run pytest packages/memtomem/tests/test_lazy_init_acceptance.py \
              packages/memtomem/tests/test_session_tracing.py -q
# 4 failed — remove either file from the invocation and everything passes
```

## Root cause (verified, three interacting parts)

1. **Lifespan tests source the developer's real `.env`.** `server/lifespan.py:108` calls `_load_dotenv()` (`lifespan.py:74-78`), which runs python-dotenv's `load_dotenv()` — resolving the repo-root `.env` from CWD. Any test that drives the real `app_lifespan` therefore imports the developer's personal env file **into `os.environ` of the whole pytest process**, outside any monkeypatch, permanently for the run. Verified polluters (empirical pair matrix, each paired with `test_session_tracing.py`):
   - `test_lazy_init_acceptance.py` — **pollutes** (real `app_lifespan`)
   - `test_server_lifespan.py` — **pollutes** (real `app_lifespan`)
   - `test_server_app_context.py`, `test_tool_registration.py` — clean (don't enter the real lifespan)

   An in-process before/after diff of `os.environ` across `pytest.main(["tests/test_lazy_init_acceptance.py"])` shows the `.env`'s keys added (`LANGFUSE_PUBLIC_KEY`, `LANGFUSE_SECRET_KEY`, `LANGFUSE_BASE_URL`, several unrelated provider API keys — values withheld here for obvious reasons).

2. **`SessionTraceConfig` reads bare, unprefixed env vars.** It subclasses `BaseSettings` with no `model_config` (`config.py:799`), so pydantic-settings binds fields to case-insensitive **unprefixed** env names: the leaked `LANGFUSE_PUBLIC_KEY`/`LANGFUSE_SECRET_KEY` populate `langfuse_public_key`/`langfuse_secret_key` on every subsequent construction — including `Mem2MemConfig()`'s nested default. Verified directly:

   ```python
   # LANGFUSE_PUBLIC_KEY=pk-test LANGFUSE_SECRET_KEY=sk-test
   SessionTraceConfig(enabled=True, langfuse_enabled=True)  # NO raise; keys filled from env
   ```

3. **The victim tests rely on keys being absent.** With both keys env-filled, `_require_keys_when_enabled` passes (and `langfuse` 4.x is installed via the extra, so `_require_langfuse_package_when_enabled` passes too) — enabling `langfuse_enabled` validates cleanly and the expected rejection/rollback never happens. Consistency check: `test_save_config_validation_rejection` still passes in full runs because it *explicitly assigns* `langfuse_public_key = ""` — assignment beats env.

## Fix options

- **A (leaning, validated) — never source the developer's `.env` from tests.** Add an autouse fixture in `tests/conftest.py` that monkeypatches `memtomem.server.lifespan._load_dotenv` to a no-op for the whole suite. Validated empirically: with exactly that fixture injected as a plugin, `test_lazy_init_acceptance.py + test_server_lifespan.py + test_session_tracing.py` all pass together (39 passed). A dedicated test can still exercise `_load_dotenv` explicitly against a tmp-CWD `.env` if we want coverage of the loader itself. This also stops tests from silently inheriting the developer's other `.env` credentials (`OPENAI_API_KEY` etc.) — a hermeticity hazard well beyond these four tests.
- **B — harden the victims only:** `monkeypatch.delenv("LANGFUSE_PUBLIC_KEY"/"LANGFUSE_SECRET_KEY")` in the four tests. Whack-a-mole: every future test touching any of the 22 unprefixed `BaseSettings` sub-configs stays exposed to whatever the `.env` happens to contain.

**Out-of-scope observation (likely its own decision):** all 22 sub-configs in `config.py` are unprefixed `BaseSettings` (`EmbeddingConfig`, `LLMConfig`, …), so generic bare env names (`API_KEY`, `ENABLED`, `MODEL`, …) bind to config fields case-insensitively. That's a footgun independent of dotenv — e.g. a shell with `ENABLED=true` exported flips `SessionTraceConfig.enabled`. Whether sub-configs should be `BaseModel` (env binding already flows through `Mem2MemConfig`'s `MEMTOMEM_` prefix + `__` nesting) deserves a separate pass; not required to fix this bug.

## Environment

macOS (Darwin 25.5.0), Python 3.12/3.13 venv, `origin/main @ cd0ae767`, `langfuse` 4.12.0 installed via the `[all]` extra. CI green on the same commit (no `.env` on runners).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

--- a/packages/memtomem/tests/conftest.py
+++ b/packages/memtomem/tests/conftest.py
@@ -2,6 +2,7 @@
 
 from __future__ import annotations
 
+import os
 import sys
 from pathlib import Path
 
@@ -19,6 +20,52 @@ from memtomem.server.component_factory import Components, create_components, clo
 # Keeping the definitions in ``_wiki_fixtures.py`` avoids bloating this
 # already-heavy conftest with unrelated git-env plumbing.
 from _wiki_fixtures import git_identity, wiki_root  # noqa: F401
+
+
+# Bare (unprefixed) env names that pydantic-settings binds onto
+# ``SessionTraceConfig`` fields case-insensitively — the sub-configs carry no
+# ``model_config``, so these bypass the documented ``MEMTOMEM_`` surface
+# entirely (#1522). A developer shell or a sourced repo-root ``.env``
+# exporting them defeats the langfuse keys-missing validator in any test
+# that constructs the config (#1508).
+_BARE_LANGFUSE_ENV_VARS = (
+    "LANGFUSE_PUBLIC_KEY",
+    "LANGFUSE_SECRET_KEY",
+    "LANGFUSE_HOST",
+    "LANGFUSE_ENABLED",
+)
+
+# The documented nested surface for the same fields. Scrubbed by prefix, not a
+# hand-list, so future SessionTraceConfig fields are covered automatically.
+_SESSION_TRACE_ENV_PREFIX = "MEMTOMEM_SESSION_TRACE__"
+
+
+@pytest.fixture(autouse=True)
+def _hermetic_dotenv(request: pytest.FixtureRequest, monkeypatch: pytest.MonkeyPatch) -> None:
+    """Keep the developer's shell and repo-root ``.env`` out of every test.
+
+    ``app_lifespan`` sources ``.env`` via ``_load_dotenv()``; python-dotenv
+    resolves the file from the *calling frame's* location (walking up from
+    ``server/lifespan.py`` to the repo root), so neither ``monkeypatch`` nor
+    ``chdir`` contains it — one lifespan-driving test loads the developer's
+    real credentials into ``os.environ`` for the rest of the pytest process
+    (#1508). Tests that need the real loader opt out with
+    ``@pytest.mark.real_dotenv``; the env scrub always applies.
+
+    The scrub is case-insensitive (pydantic-settings binds env names
+    case-insensitively, so lowercase ``langfuse_public_key`` leaks the same
+    way) and also covers the documented ``MEMTOMEM_SESSION_TRACE__*`` surface
+    a developer shell may export. Tests that need any of these set them via
+    ``monkeypatch.setenv`` in their own body, which runs after this fixture.
+    """
+    for key in list(os.environ):
+        upper = key.upper()
+        if upper in _BARE_LANGFUSE_ENV_VARS or upper.startswith(_SESSION_TRACE_ENV_PREFIX):
+            monkeypatch.delenv(key, raising=False)
+    if request.node.get_closest_marker("real_dotenv") is None:
+        from memtomem.server import lifespan as lifespan_mod
+
+        monkeypatch.setattr(lifespan_mod, "_load_dotenv", lambda: None)
 
 
 def _ollama_available() -> bool:

--- a/packages/memtomem/tests/test_server_lifespan.py
+++ b/packages/memtomem/tests/test_server_lifespan.py
@@ -16,6 +16,7 @@ are covered in ``test_server_app_context.py``.
 
 from __future__ import annotations
 
+import os
 from unittest.mock import AsyncMock, MagicMock
 
 import pytest
@@ -217,3 +218,65 @@ async def test_lifespan_cleans_up_webhook_when_appcontext_init_fails(
             pytest.fail("yield should not be reached")
 
     assert closed, "webhook must be closed when AppContext construction fails"
+
+
+# ── dotenv loading (#1508) ───────────────────────────────────────────
+
+
+@pytest.mark.asyncio
+async def test_lifespan_invokes_dotenv_loader(monkeypatch: pytest.MonkeyPatch) -> None:
+    """Startup must call ``_load_dotenv()`` exactly once. The suite-wide
+    conftest fixture no-ops the loader (#1508), so this spy — layered on
+    top of that no-op — is the only remaining coverage that the production
+    startup path still wires it."""
+    monkeypatch.delenv("MEMTOMEM_WEBHOOK__ENABLED", raising=False)
+    monkeypatch.delenv("MEMTOMEM_WEBHOOK__URL", raising=False)
+
+    calls: list[None] = []
+    monkeypatch.setattr(lifespan_mod, "_load_dotenv", lambda: calls.append(None))
+
+    async with lifespan_mod.app_lifespan(MagicMock()):
+        pass
+
+    assert len(calls) == 1, f"_load_dotenv must run once at startup; ran {len(calls)}×"
+
+
+@pytest.mark.asyncio
+async def test_lifespan_under_test_does_not_source_repo_dotenv(
+    monkeypatch: pytest.MonkeyPatch,
+) -> None:
+    """End-to-end wiring pin for #1508: with the conftest hermeticity
+    fixture active, entering the *real* lifespan must leave the bare
+    langfuse env absent — even on a dev machine whose repo-root ``.env``
+    defines it. Before the fix, this test polluted ``os.environ`` for
+    every later test in the run (the four ``test_session_tracing``
+    validator failures)."""
+    monkeypatch.delenv("MEMTOMEM_WEBHOOK__ENABLED", raising=False)
+    monkeypatch.delenv("MEMTOMEM_WEBHOOK__URL", raising=False)
+
+    async with lifespan_mod.app_lifespan(MagicMock()):
+        pass
+
+    assert "LANGFUSE_PUBLIC_KEY" not in os.environ
+    assert "LANGFUSE_SECRET_KEY" not in os.environ
+
+
+@pytest.mark.real_dotenv
+def test_load_dotenv_does_not_override_existing_env(
+    monkeypatch: pytest.MonkeyPatch,
+) -> None:
+    """The real loader must keep already-exported env vars (python-dotenv's
+    ``override=False`` default): a repo-root ``.env`` may *add* to the
+    environment but never replace explicit configuration. Marked
+    ``real_dotenv`` to opt out of the conftest no-op; the loader resolves
+    ``.env`` upward from the source tree, so on a dev machine it may add
+    that file's other keys — the snapshot/restore keeps this test from
+    becoming the very polluter #1508 fixed."""
+    snapshot = dict(os.environ)
+    monkeypatch.setenv("LANGFUSE_PUBLIC_KEY", "sentinel-keep-me")
+    try:
+        lifespan_mod._load_dotenv()
+        assert os.environ["LANGFUSE_PUBLIC_KEY"] == "sentinel-keep-me"
+    finally:
+        os.environ.clear()
+        os.environ.update(snapshot)

--- a/packages/memtomem/tests/test_session_cli.py
+++ b/packages/memtomem/tests/test_session_cli.py
@@ -5,6 +5,7 @@ from __future__ import annotations
 
 import json
 from contextlib import asynccontextmanager
+from pathlib import Path
 from types import SimpleNamespace
 from unittest.mock import AsyncMock
 
@@ -12,6 +13,8 @@ import pytest
 from click.testing import CliRunner
 
 from memtomem.cli import cli
+
+from .helpers import set_home
 
 
 def _mock_components(*, sessions=None, events=None, add_event=None):
@@ -51,7 +54,12 @@ def _create_session_call_args(mock: AsyncMock) -> tuple:
 
 
 @pytest.fixture
-def runner() -> CliRunner:
+def runner(monkeypatch: pytest.MonkeyPatch, tmp_path: Path) -> CliRunner:
+    """CliRunner with HOME isolated. The CLI bootstrap otherwise reads the
+    developer's real ``~/.memtomem/config.json``; one that enables langfuse
+    tracing (keys expected from env) makes the client's auth warning bleed
+    into ``result.output`` and break the ``--json`` contract (#1508)."""
+    set_home(monkeypatch, tmp_path)
     return CliRunner()
 
 

--- a/packages/memtomem/tests/test_web_routes_extended.py
+++ b/packages/memtomem/tests/test_web_routes_extended.py
@@ -1751,8 +1751,15 @@ class TestConfigPatch:
         assert any("nonexistent" in r for r in data["rejected"])
 
     async def test_save_config(self, client: AsyncClient):
-        """POST /api/config/save persists config."""
-        with patch("memtomem.config.save_config_overrides"):
+        """POST /api/config/save persists config.
+
+        Patch the *route module's* binding, like ``test_patch_namespace_rules``
+        below — patching ``memtomem.config.save_config_overrides`` misses the
+        ``from memtomem.config import ...`` binding in ``routes/system.py``, so
+        the real saver ran against the developer's ``~/.memtomem/config.json``
+        (masked while leaked langfuse env kept its validation green, #1508).
+        """
+        with patch("memtomem.web.routes.system.save_config_overrides"):
             resp = await client.post("/api/config/save")
             assert resp.status_code == 200
             data = resp.json()

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -42,6 +42,7 @@ markers = [
     "multilingual: multilingual/large-corpus regression tests (runs in the test-golden-path CI job)",
     "requires_symlinks: needs a filesystem that can create symlinks (auto-skipped on Windows without Developer Mode/admin)",
     "browser: requires pytest-playwright + a Chromium install (auto-skipped when unavailable)",
+    "real_dotenv: opt out of the suite-wide _load_dotenv no-op and run the real dotenv loader (#1508)",
 ]
 
 [tool.mypy]

